aptos-move/framework/aptos-framework/sources/TokenV1.move

```diff
@@ -0,0 +1,935 @@
+/// This module provides the foundation for Tokens.
+module AptosFramework::TokenV1 {
+    use Std::ASCII::{String, Self};
+    use Std::Errors;
+    use Std::Event::{Self, EventHandle};
+    use Std::Signer;
+    use Std::Vector;
+
+    use AptosFramework::Table::{Self, Table};
+    use AptosFramework::PropertyList::{Self, PropertyList};
+
+    const TOKEN_MAX_MUTABLE_IND: u64 = 0;
+    const TOKEN_URI_MUTABLE_IND: u64 = 1;
+    const TOKEN_DESCRIPTION_MUTABLE_IND: u64 = 2;
+    const TOKEN_ROYALTY_MUTABLE_IND: u64 = 3;
+    const TOKEN_PROPERTY_MUTABLE_IND: u64 = 4;
+
+    const COLLECTION_DESCRIPTION_MUTABLE_IND: u64 = 0;
+    const COLLECTION_URI_MUTABLE_IND: u64 = 1;
+    const COLLECTION_MAX_MUTABLE_IND: u64 = 2;
+
+    const EALREADY_HAS_BALANCE: u64 = 0;
+    const EBALANCE_NOT_PUBLISHED: u64 = 1;
+    const ECOLLECTIONS_NOT_PUBLISHED: u64 = 2;
+    const ECOLLECTION_NOT_PUBLISHED: u64 = 3;
+    const ECOLLECTION_ALREADY_EXISTS: u64 = 4;
+    const ECREATE_WOULD_EXCEED_MAXIMUM: u64 = 5;
+    const EINSUFFICIENT_BALANCE: u64 = 6;
+    const EINVALID_COLLECTION_NAME: u64 = 7;
+    const EINVALID_TOKEN_MERGE: u64 = 8;
+    const EMINT_WOULD_EXCEED_MAXIMUM: u64 = 9;
+    const ENO_BURN_CAPABILITY: u64 = 10;
+    const ENO_MINT_CAPABILITY: u64 = 11;
+    const ETOKEN_ALREADY_EXISTS: u64 = 12;
+    const ETOKEN_NOT_PUBLISHED: u64 = 13;
+    const ETOKEN_STORE_NOT_PUBLISHED: u64 = 14;
+
+    //
+    // Core data structures for holding tokens
+    //
+
+    struct Token has store {
+        id: TokenId,
+        value: u64,
+    }
+
+    struct TokenId has store, copy, drop {
+        tokendata_id: TokenDataId,
+        serial_number: u64
```

Comment:
I thought we wanted this to optional so that if is set to none we'd allow an amount?

Reply:
I think we discussed this previously.  Each TokenId maps a a token with an amount. When people creating the token, they can decide the amount. There was a question on if the token is still an NFT. This is decided when minting. Once minting an serial_number with an amount of 1 and the tokendata is not mutable, the token is unique and it is an NFT.
 
Every mint will increase the serial_number.